metals/src/main/scala/scala/meta/internal/metals/ProjectMetalsLspService.scala

```diff
@@ -436,6 +441,34 @@ class ProjectMetalsLspService(
     }
   }
 
+  private val previousCreateOrModify = TrieMap[AbsolutePath, String]()
+
+  // If code is genereted we might get a lot of `CreateOrModify`
+  // events, even if the code doesn't really change
+  private def didContentChange(event: FileWatcherEvent): Boolean = {
+    val path = AbsolutePath(event.path)
+    event.eventType match {
+      case EventType.CreateOrModify =>
+        try {
+          val md5 = MD5.bytesToHex(FileIO.readAllBytes(path))
```

Comment:
Maybe we should do it before compile? So that we wouldn't do noop compilations as well?

Reply:
I put it before compile. I also added a draft of a solution to deduplicate reading in file content. (I feel like that might be costly.) WDYT?

Reply:
I realised we also compile the file when it did not change but its dependencies did (on `didFocus`). I wonder if it's safe to always discard compile on `didSave` if the file did not change. There may be some compilation that finished and user may want to force compile using save. Maybe the first option to only deduplicate on file watcher events was better?

Reply:
> I realised we also compile the file when it did not change but its dependencies did (on didFocus). I wonder if it's safe to always discard compile on didSave if the file did not change.

But if they want to force a save with didSave, they need to change the file, right? Did save will not be sent on a unchanged file anyway.

If any upstream projects change, we should always maybe mark anything that depends on them as "dirty" and should always be recompiled next time a compilation is invoked?

Reply:
> But if they want to force a save with didSave, they need to change the file, right? Did save will not be sent on a unchanged file anyway

At least in VSCode `Save` will always send `didSave` even if the file did not change.

Reply:
I don't see didSave being sent if I save an unchanged file :thinking: Or do you mean in a different case?